spring-pulsar/src/test/java/org/springframework/pulsar/reader/DefaultPulsarReaderListenerContainerTests.java

```diff
@@ -0,0 +1,155 @@
+ * limitations under the License.
+ */
+
+package org.springframework.pulsar.reader;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.pulsar.client.api.MessageId;
+import org.apache.pulsar.client.api.PulsarClient;
+import org.apache.pulsar.client.api.PulsarClientException;
+import org.apache.pulsar.client.api.ReaderListener;
+import org.apache.pulsar.client.api.Schema;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import org.springframework.lang.Nullable;
+import org.springframework.pulsar.core.DefaultPulsarProducerFactory;
+import org.springframework.pulsar.core.DefaultPulsarReaderFactory;
+import org.springframework.pulsar.core.PulsarTemplate;
+import org.springframework.pulsar.test.support.PulsarTestContainerSupport;
+
+/**
+ * Basic tests for {@link DefaultPulsarReaderListenerContainer}.
+ *
+ * @author Soby Chacko
+ */
+public class DefaultPulsarReaderListenerContainerTests implements PulsarTestContainerSupport {
+
+	@Nullable
+	private PulsarClient pulsarClient;
+
+	@BeforeEach
+	void createPulsarClient() throws PulsarClientException {
+		pulsarClient = PulsarClient.builder().serviceUrl(PulsarTestContainerSupport.getPulsarBrokerUrl()).build();
+	}
+
+	@AfterEach
+	void closePulsarClient() throws PulsarClientException {
+		if (pulsarClient != null && !pulsarClient.isClosed()) {
+			pulsarClient.close();
+		}
+	}
+
+	@Test
+	void basicDefaultReader() throws Exception {
+		Map<String, Object> config = Map.of("topicNames", Collections.singleton("dprlct-001"), "subscriptionName",
+				"dprlct-sub-001");
+
+		DefaultPulsarReaderFactory<String> pulsarReaderFactory = new DefaultPulsarReaderFactory<>(pulsarClient, config);
+		CountDownLatch latch = new CountDownLatch(1);
+		PulsarReaderContainerProperties readerContainerProperties = new PulsarReaderContainerProperties();
+		readerContainerProperties.setReaderListener((ReaderListener<?>) (reader, msg) -> {
+			assertThat(msg.getValue()).isEqualTo("hello john doe");
+			latch.countDown();
+		});
+		readerContainerProperties.setStartMessageId(MessageId.earliest);
+		readerContainerProperties.setSchema(Schema.STRING);
+		DefaultPulsarReaderListenerContainer<String> container = new DefaultPulsarReaderListenerContainer<>(
+				pulsarReaderFactory, readerContainerProperties);
+		container.start();
+
+		Map<String, Object> prodConfig = Map.of("topicName", "dprlct-001");
+		DefaultPulsarProducerFactory<String> pulsarProducerFactory = new DefaultPulsarProducerFactory<>(pulsarClient,
+				prodConfig);
+		PulsarTemplate<String> pulsarTemplate = new PulsarTemplate<>(pulsarProducerFactory);
+		pulsarTemplate.sendAsync("hello john doe");
+		assertThat(latch.await(10, TimeUnit.SECONDS)).isTrue();
+		container.stop();
```

Comment:
I think we should start this test out early on by always stopping the container in a finally block. I reworked this test to use `var` and `try-finally` and it looks like the following. I think it is still clean w/ the try/finally. 

```java
	@Test
	void latestMessageId() throws Exception {
		var latch = new CountDownLatch(2);
		var containerProps = new PulsarReaderContainerProperties();
		containerProps.setReaderListener((ReaderListener<?>) (reader, msg) -> {
			assertThat(msg.getValue()).isEqualTo("This message should be received by the reader");
			latch.countDown();
		});
		containerProps.setStartMessageId(MessageId.latest);
		containerProps.setTopics(List.of("dprlct-003"));
		containerProps.setSchema(Schema.STRING);

		var readerConfig = Collections.<String, Object>emptyMap();
		var readerFactory = new DefaultPulsarReaderFactory<String>(pulsarClient, readerConfig);
		DefaultPulsarReaderListenerContainer<String> container = null;
		try {
			container = new DefaultPulsarReaderListenerContainer<>(readerFactory, containerProps);
			
			var prodConfig = Map.<String, Object>of("topicName", "dprlct-003");
			var producerFactory = new DefaultPulsarProducerFactory<>(pulsarClient, prodConfig);
			var pulsarTemplate = new PulsarTemplate<>(producerFactory);

			// The following sends will not be received by the reader as we are using the
			// latest message id to start from.
			for (int i = 0; i < 5; i++) {
				pulsarTemplate.sendAsync("This message should not be received by the reader");
			}
			container.start();

			pulsarTemplate.sendAsync("This message should be received by the reader");
			pulsarTemplate.sendAsync("This message should be received by the reader");

			assertThat(latch.await(10, TimeUnit.SECONDS)).isTrue();
		}
		finally {
			safeStopContainer(container);
		}
	}
```